### ISSUE TITLE
NewlineAtEndOfFileCheck and empty files (size 0)

### ISSUE THREAD

**Issue by alexandrul** [checkstyle/checkstyle#17250]
I have read check documentation: https://checkstyle.org/checks/xxxxxx/nameofaffectedcheck.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
/var/tmp $ cat config.xml
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="NewlineAtEndOfFile"/>
</module>

/var/tmp $ RUN_LOCALE="-Duser.language=en -Duser.country=US"
/var/tmp $ java $RUN_LOCALE -jar checkstyle-10.25.0-all.jar -c config.xml checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/newlineatendoffile/InputNewlineAtEndOfFileEmptyFile.txt

Starting audit...
[ERROR] /home/me/java/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/newlineatendoffile/InputNewlineAtEndOfFileEmptyFile.txt:1: File does not end with a newline. [NewlineAtEndOfFile]
Audit done.
Checkstyle ends with 1 errors.
```

---

`NewlineAtEndOfFile` is reporting files with size 0 as errors.
The current workaround is to add a single line separator to the file.

Is it possible to revert this decision?

**Comment by romani** [comment]
Please share a reason why you need empty files

**Comment by alexandrul** [comment]
I don't need them in my personal projects.

At work, the files that are preventing me to enable this check for a legacy project are not under my control.

**Comment by romani** [comment]
What is incentive for your to have this fix?
If project is so legacy, you can use suppressions to skip dealing with them.

I can probably understand why non java file can be empty in git repository - to keep folder in git.
But why empty java file might be required?

We need reason, to share it in docs, as someone can come to us with conflicting requests to put violations on empty files.

**Comment by romani** [comment]
I am ok to exclude empty files from this Check validation.
In rationale of this Check, goal is clear why this Check matter, but empty file is ok to be skipped by this Check.

Benefits of this Check is reduction of diff on next update. But from nothing to some content, diff will be same. So Check is not beneficial on such cases.

If users need empty file as marker, adding symbol to file will make it as not empty and it will be more confusing for users.

@nrmancuso , please review this issue and put label if you agree.

**Comment by romani** [comment]
@rdiachenko , please review this issue also.

**Comment by nrmancuso** [comment]
I would rather create some xpath suppression that can be easily reused.

**Comment by rdiachenko** [comment]
Empty files are usually some markers for other systems. Different tools may generate them and it may be a pain to always update them by inserting a new line. I agree some generic suppression may cover this, but this looks like a common edge case to me, wich makes sense to extend this check with an option to ignore empty files (turned off by default).

**Comment by alexandrul** [comment]
Regardless of its reason of being, an empty file:
- it won't receive a line separator on save in VS Code (with `"files.insertFinalNewline": true`)
- it won't receive a line separator on save in IntelliJ and PyCharm (with `Ensure every saved file ends with a line break` enabled)
- is not reported by the `Files should contain an empty newline at the end` SonarQube rule

While it would be great to change the default behavior, an option to ignore empty files would also be fine.

As a side note, is it worth adding a new check for empty files?
`NewlineAtEndOfFileCheck` is not an obvious option when looking for empty files checks.

**Comment by romani** [comment]
After more thinking on this, I am approving issue, PR is welcome.

**Comment by romani** [comment]
@alexandrul , do you prefer release sooner ( today , tomorrow) our you can wait until end of week?

**Comment by alexandrul** [comment]
@romani I can wait as long as necessary, even a couple of months.

**Comment by romani** [comment]
We encourage user to send us fixes and we can deploy right after merge to keep athor of fix very incentified.  Ok we will release on weekend by schedule.
